crates/component/src/wrappers/http.rs

```diff
@@ -0,0 +1,422 @@
+//! This module provides utilities for writing HTTP servers and clients using the WASI HTTP API.
+//!
+//! It's inspired by the WASI 0.3 proposal for <https://github.com/WebAssembly/wasi-http> and will
+//! be supported until the release of wasi:http@0.3.0. After that, this module will be deprecated.
+//!
+//! This crate can be used with the [reqwest-wasmcloud](https://github.com/wasmCloud/reqwest-wasmcloud)
+//! crate to make outgoing HTTP requests from a WebAssembly component.
+//!
+//! ```rust
+//! use wasi::http::types::ErrorCode;
+//! use wasmcloud_component::http::{HttpServer, Request, ResponseBuilder};
+//!
+//! struct Component;
+//!
+//! wasmcloud_component::http::export!(Component);
+//!
+//! // Implementing the [`HttpServer`] trait for a component
+//! impl HttpServer for Component {
+//!     fn handle(_request: Request) -> Result<ResponseBuilder, ErrorCode> {
+//!         Ok(ResponseBuilder::ok("Hello from Rust!"))
+//!     }
+//! }
+//! ```
+// Required until <https://github.com/seanmonstar/reqwest/pull/2453> merges
+use reqwest_wasmcloud as reqwest;
+
+use std::{
+    io::{Read, Write},
+    str::FromStr,
+};
+
+use reqwest::header::{HeaderMap, HeaderName, HeaderValue};
+use wasi::{
+    http::{
+        outgoing_handler::ErrorCode,
+        types::{
+            Fields, IncomingBody, InputStream, Method, OutgoingBody, OutgoingResponse,
+            ResponseOutparam, Scheme,
+        },
+    },
+    io::streams::StreamError,
+};
+
+const HTTP_SCHEME: &str = "http";
+const HTTPS_SCHEME: &str = "https";
+
+/// Trait for implementing an HTTP server WebAssembly component that receives a
+/// [`Request`] and returns a [`ResponseBuilder`].
+pub trait HttpServer {
+    fn handle(request: Request) -> Result<ResponseBuilder, ErrorCode>;
```

Comment:
I'm surprised this returns a `ResponseBuilder`, rather than a complete Response

I'd expect to be able to receive `http::Request`(https://docs.rs/http/latest/http/request/struct.Request.html) here and return an `http::Response` (https://docs.rs/http/latest/http/response/struct.Response.html)

I'd also expect to be able to set an arbitrary `http_body::Body` as the body of the response https://docs.rs/http-body/latest/http_body/trait.Body.html

Reply:
Thank you for the example, let me see what I can do to refactor this a bit. Returning the builder _is_ weird, I believe I did it this way so that I could deal with ownership of the response resources properly but I think I can refactor to return a response.

The only reason I have `wasmcloud_component::http::Response` is so that I can handle a `Vec<u8>` body and a `InputStream` body, so I might need to go back to the drawing board there. It would be great to use `http`